src/Features/Core/Portable/Diagnostics/DiagnosticHelper.cs

```diff
@@ -55,6 +58,44 @@ internal static class DiagnosticHelper
             return CreateWithMessage(descriptor, location, effectiveSeverity, additionalLocations, properties, message);
         }
 
+        /// <summary>
+        /// Create a diagnostic that adds properties specifying a tag for a set of locations.
+        /// </summary>
+        /// <param name="descriptor">A <see cref="DiagnosticDescriptor"/> describing the diagnostic.</param>
+        /// <param name="location">An optional primary location of the diagnostic. If null, <see cref="Location"/> will return <see cref="Location.None"/>.</param>
+        /// <param name="effectiveSeverity">Effective severity of the diagnostic.</param>
+        /// <param name="additionalLocations">
+        /// An optional set of additional locations related to the diagnostic.
+        /// Typically, these are locations of other items referenced in the message.
+        /// If null, <see cref="Diagnostic.AdditionalLocations"/> will return an empty list.
+        /// </param>
+        /// <param name="tagIndices">a map of location tag to index in additional locations.</param>
+        /// <param name="messageArgs">Arguments to the message of the diagnostic.</param>
+        /// <returns>The <see cref="Diagnostic"/> instance.</returns>
+        public static Diagnostic CreateWithLocationTags(
+            DiagnosticDescriptor descriptor,
+            Location location,
+            ReportDiagnostic effectiveSeverity,
+            IEnumerable<Location> additionalLocations,
+            IDictionary<string, IEnumerable<int>> tagIndices,
+            params object[] messageArgs)
+        {
```

Comment:
Consider adding an assert here about non-zero additional locations, non-zero tag indices, and the fact that max of all indices must be less than or equals additional location count. Also consider adding an example in the comment from your analyzer that is currently using it.